### ISSUE TITLE
compile straight to jar

### ISSUE THREAD

**Issue by fommil** [sbt/sbt#2293]
On Windows, there is a significant overhead to having lots of files and even the act of reading/writing classfiles can cost 10% of the project compile time. This is a limitation of the NTFS filesystem.

It would be incredibly useful if there were an option to compile straight to jar as such access is significantly faster.

Note that this would simplify the implementation of #2270

**Comment by fommil** [comment]
@adriaanm I spoke to you briefly about this idea at Scala World, but I forgot to file it at the time. Possibly requires `scalac` support?

**Comment by adriaanm** [comment]
Yep, this would require compiler support. There may already be some support there -- not sure what the status is off the top of my head.

**Comment by fommil** [comment]
It'd be great to get some performance stats on this, just to understand what the gains could be.

**Comment by fommil** [comment]
A proof point could be simulating writing a million small files out to disk, vs writing them out to 100 jars.

Would give us an idea what the gains could be.

**Comment by adriaanm** [comment]
Now that I have a real computer in front of me, I managed to dig up what I was remembering: https://github.com/scala/scala/commit/2700617052

**Comment by fommil** [comment]
@paulp did you test that on Windows?

**Comment by adriaanm** [comment]
Good one :)

**Comment by fommil** [comment]
I try

**Comment by paulp** [comment]
I distinctly remember an animated discussion amongst the Windows team about the finer points of writing to the windows filesystem. Somebody told a joke about a bunny and a circus clown walking into a bar. We all had a good laugh. I'm sure they tested it thoroughly.

**Comment by paulp** [comment]
Incidentally, the reason one never sees tests for this sort of thing is spelled W-I-N-D-O-W-S. It's known as the law of perverse incentive reversal.

**Comment by paulp** [comment]
On the plus side, looks like it still works otherwise.

**Comment by fommil** [comment]
All the Windows-specific functions in emacs start with `w32-` because they couldn't quite bring themselves to start them `win-`.

I might knock up a speed test as described above at some point to see what the potential gain could be.

**Comment by alexarchambault** [comment]
About gains on Windows when generating less class files, I definitely saw some when I was using Windows some months ago, in the context of `shapeless.Lazy` (but didn't take the time to setup benchmarks then), like this guy did: https://gitter.im/milessabin/shapeless?at=55d304f8057d8c9d3a6ddf44 (and read down to [this message](https://gitter.im/milessabin/shapeless?at=55d30f6a255950880cfc04b8) to correct some intermediary messages).

**Comment by paulp** [comment]
For what it's worth, on OSX compiling 220K lines of code to dir vs. jar, I get 17.1 seconds spent in phase jvm when writing to a directory and 14.6 seconds writing to a jar. I/O being I/O and laptops being laptops this difference might be noise, but it feels about right.

**Comment by fommil** [comment]
yeah, but that's almost a real operating system.

**Comment by fommil** [comment]
TL;DR this is definitely worth looking at!

some test code and numbers (for 10,000 files on Windows 7):

```
[info] Running ClassVsJar 
[info] Starting classfile test
[info] Creating classfiles took: 5.049 s
[info] Bonus statistic: deleting classfiles took 5.291 s
[info] Starting jar test
[info] Creating jar (compress = 9) took: 7.873 s
[info] Bonus statistic: deleting jar took 553.0 ?s
[info] Starting jar test
[info] Creating jar (compress = 8) took: 3.198 s
[info] Bonus statistic: deleting jar took 531.9 ?s
[info] Starting jar test
[info] Creating jar (compress = 7) took: 3.133 s
[info] Bonus statistic: deleting jar took 550.3 ?s
[info] Starting jar test
[info] Creating jar (compress = 6) took: 3.126 s
[info] Bonus statistic: deleting jar took 759.2 ?s
[info] Starting jar test
[info] Creating jar (compress = 5) took: 3.448 s
[info] Bonus statistic: deleting jar took 583.5 ?s
[info] Starting jar test
[info] Creating jar (compress = 4) took: 3.458 s
[info] Bonus statistic: deleting jar took 516.5 ?s
[info] Starting jar test
[info] Creating jar (compress = 3) took: 2.878 s
[info] Bonus statistic: deleting jar took 473.3 ?s
[info] Starting jar test
[info] Creating jar (compress = 2) took: 2.828 s
[info] Bonus statistic: deleting jar took 672.5 ?s
[info] Starting jar test
[info] Creating jar (compress = 1) took: 2.801 s
[info] Bonus statistic: deleting jar took 648.1 ?s
[info] Starting jar test
[info] Creating jar (compress = 0) took: 3.734 s
[info] Bonus statistic: deleting jar took 4.624 success
```

``` scala
import java.io._
import java.util.jar._

import com.google.common.base.Stopwatch
import org.apache.commons.io.{ FileUtils, IOUtils }

object ClassVsJar extends App {

  val dir = new File("tmp-dir")
  dir.mkdirs()

  val jarFile = new File("tmp.jar")

  // factor out the cost of generating this
  val names = (1 to 10000).map(_.toString)

  // always write out this
  val contents = ("foobar" * 1000).getBytes

  try {
    println("Starting classfile test")
    val watch = Stopwatch.createStarted()

    names.foreach { name =>
      FileUtils.writeByteArrayToFile(new File(dir, name), contents)
    }

    println("Creating classfiles took: " + watch)
  } finally {
    val watch = Stopwatch.createStarted()
    FileUtils.deleteDirectory(dir)
    println("Bonus statistic: deleting classfiles took " + watch)
  }

  // there is some overhead to instantiating the machinery
  (9 to 0 by -1).foreach { level =>
    val jar = new JarOutputStream(new FileOutputStream(jarFile))
    jar.setLevel(level)
    try {
      println("Starting jar test")
      val watch = Stopwatch.createStarted()
      names.foreach { name =>

        // less restrictions on entry name length!
        val entry = new JarEntry(name)
        jar.putNextEntry(entry)

        IOUtils.copy(new ByteArrayInputStream(contents), jar)

        jar.closeEntry()
      }

      jar.close()

      println(s"Creating jar (compress = $level) took: " + watch)
    } finally {
      val watch = Stopwatch.createStarted()
      jarFile.delete()
      println("Bonus statistic: deleting jar took " + watch)
    }
  }
}
```

**Comment by eed3si9n** [comment]
If `DEFAULT_COMPRESSION` is [6](http://stackoverflow.com/questions/16485573/what-exactly-is-default-compression), then we're talking about 3.126s overhead for 10k files. That doesn't sound all that bad, unless you're doing it _all the time_.

If this is stemming from `exportJars := true` causing useless JAR making, which is being discussed in #2270, then caching of `packageBin` should be the direction we should explore.

**Comment by fommil** [comment]
I'm not really concerned about the overhead of the jar creation (although, yeah, probably we should use `1` for a quick win), but the overhead of the `classfile` creation is **huge** so this ticket is really about compiling straight into a `jar` and never writing out the intermediate `.class` files.

**Comment by fommil** [comment]
@soc

**Comment by fommil** [comment]
@eed3si9n you said this would involve a big refactor of zinc so unlikely until another major release (I'm guessing post sbt-1.0) but what about an sbt `Task` to compile straight to jar for a full module without using incremental compile? That might be really useful for situations where you have a clean checkout and want to just do a full compile. Or for CI jobs where there is no incremental compilation.

**Comment by eed3si9n** [comment]
If you just want a full compilation task that might be doable.

**Comment by fommil** [comment]
:+1: I'm all about the tactical strike :grinning:

**Comment by fommil** [comment]
(important thing is to skip the creation/deletion of class files on Windows wherever possible)

**Comment by cunei** [comment]
I put together a plugin that should address this issue. See: https://github.com/typesafehub/sbt-tojar
The files should be available on Bintray shortly for initial testing; please note that it requires a not-yet-released 0.13.10-RC2 version of sbt.

**Comment by dwijnand** [comment]
Really nice work there..

**Comment by fommil** [comment]
great! I'll try to have a look shortly.
